### DISCUSSION TITLE
Speed up the build process

### DISCUSSION THREAD

**Discussion by Ho-Ro** [avrdudes/avrdude#1585] [Ideas]
The build process from `build.sh` uses only one processor core for make.
With the parameter `-j4` or even more aggressive `-j` a significant speed up can be achieved, even on my old development ThinkPad T400 with 2 cores, where I normally use `make -j4`. 
The modification should work on all Unixoids in use.

```diff
--- build.sh    2023-09-23 16:45:14.853977259 +0200
+++ build_fast.sh       2023-12-01 09:37:17.300721493 +0100
@@ -103,7 +103,7 @@
 cd build_${ostype}
 cmake ${build_flags} ${extra_enable} -D CMAKE_BUILD_TYPE=${build_type} .. ||\
     { echo "CMake failed."; exit 1; }
-cmake --build . ||\
+cmake --build . -- -j4 ||\
     { echo "Build failed."; exit 1; }
 
 cat <<EOF
```

Standard `build.sh`:
```
real    0m44,509s
user    0m39,936s
sys     0m4,148s
```

Modified `build_fast.sh`:
```
real    0m27,738s
user    0m44,949s
sys     0m4,130s
```

**Comment by dl8dtl** [comment]
While I understand the intention, I'm not so sure whether hardcoding a number of jobs is a good idea. Imagine someone would want to run it inside a VM that has only been granted one CPU, just for one example.

Ideally, we find a way to figure out the number of CPUs / cores, but I'm afraid there's no good portable way for that. But at the least, a commandline option or environment variable should be there (and documented) that can override the default.

**Comment by Ho-Ro** [comment]
> I'm not so sure whether hardcoding a number of jobs is a good idea. Imagine someone would want to run it inside a VM that has only been granted one CPU, just for one example.

Jörg, I disagree, because `-j` doesn't specify the number of cores it must use, but the number of make jobs running in parallel, so if only one core is available (or granted), it only uses that core. Also on a one-core machine you have many different processes running in parallel.
This `-j` option has long been part of `make`, even in the days when hardly any computer had more than one core, even in the days when hardly any computer had more than one core -  even when I built my first 1.xx kernels, I typed `make -j2` or `-j4` just to avoid CPU idle when IO lags occurred (just `-j` alone was a bad thing though, because it created hundreds of kernel build processes on a 8 MB machine).
What I see is that only with `make` one of my two cores idles alternately. Building `avrdude` is a straightforward process with no major dependencies between the make results, requiring no wait times between compiles except for the library build.

```
$man make
...
-j [jobs], --jobs[=jobs]
    Specifies  the number of jobs (commands) to run simultaneously.  If there is more than one -j option, the
    last one is effective.  If the -j option is given without an argument, make will not limit the number  of
    jobs  that can run simultaneously. When make invokes a sub-make, all instances of make will coordinate to
    run the specified number of jobs at a time; see the section PARALLEL MAKE AND THE JOBSERVER for details.
```

**Comment by dl8dtl** [comment]
I know what `-j <N>` means. Back in those old days, you usually had to wait for the disk to deliver or write data. All this is a lot faster now, so trying to use more jobs in parallel than you have CPUs is mostly not a good idea, as it creates additional bottlenecks.
That's why I would like to have it configurable.

**Comment by Ho-Ro** [comment]
Please check #1586

**Comment by mcuee** [comment]
#1586 has been merged.
* #1586
